File: app/src/main/java/com/juniperphoton/myersplash/widget/ProgressView.java
package com.juniperphoton.myersplash.widget;


import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import com.juniperphoton.myersplash.R;

public class ProgressView extends View {

    private int mColor;
    private int mProgress;

    public ProgressView(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.ProgressView, 0, 0);
        mColor = array.getInt(R.styleable.ProgressView_BackgroundColor, Color.WHITE);
        mProgress = array.getInt(R.styleable.ProgressView_Progress, 0);
        array.recycle();
    }

    public void setProgress(int progress) {
        mProgress = progress;
        invalidate();
    }

    public void setThemeColor(int color) {
        mColor = color;
    }

    public void animateProgressTo(int progress) {
        ValueAnimator animator = ValueAnimator.ofInt(mProgress, progress);
        animator.setDuration(300);
        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                setProgress((int) valueAnimator.getAnimatedValue());
            }
        });
        animator.start();
    }

    @Override
    public void draw(Canvas canvas) {
        super.draw(canvas);

        Paint paint = new Paint();
        paint.setColor(mColor);

        int width = getWidth();
        int progressWidth = (int) (width * (mProgress / 100d));

        canvas.drawRect(0, 0, progressWidth, getHeight(), paint);
    }
}

File: app/src/main/kotlin/com/juniperphoton/myersplash/widget/ProgressView.kt
package com.juniperphoton.myersplash.widget

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View

import com.juniperphoton.myersplash.R

class ProgressView(context: Context, attrs: AttributeSet) : View(context, attrs) {
    var color: Int = 0
    var progress: Int = 0
        set(value) {
            field = value
            invalidate()
        }

    init {
        val array = context.obtainStyledAttributes(attrs, R.styleable.CustomProgressView, 0, 0)
        color = array.getInt(R.styleable.CustomProgressView_background_color, Color.WHITE)
        progress = array.getInt(R.styleable.CustomProgressView_custom_progress, 0)
        array.recycle()
    }

    fun animateProgressTo(progress: Int) {
        val animator = ValueAnimator.ofInt(this.progress, progress)
        animator.duration = 300
        animator.addUpdateListener { valueAnimator -> this.progress = (valueAnimator.animatedValue as Int) }
        animator.start()
    }

    override fun draw(canvas: Canvas) {
        super.draw(canvas)

        val paint = Paint()
        paint.color = color

        val width = width
        val progressWidth = (width * (progress / 100.0)).toInt()

        canvas.drawRect(0f, 0f, progressWidth.toFloat(), height.toFloat(), paint)
    }
}
